hedera-node/docs/design/services/smart-contract-service/update_token_custom_fees_sc.md

```diff
@@ -0,0 +1,87 @@
+[HIP-514](https://hips.hedera.com/hip/hip-514) - HIP that introduces token management via smart contracts.
+
+## Goals
+
+- Expose `updateFungibleTokenCustomFees` and `updateNonFungibleTokenCustomFees` as new functions in the Hedera Token Service Smart Contract.
+- Implement the needed HTS system contract classes to support the new function.
+
+## Non Goals
+
+- The implementation of the HAPI operation, as it is already an existing feature.
+
+## Architecture
+
+The architecture for update token custom fees follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.
+
+## Implementation
+
+### New Function Implementations
+
+New system contract functions must be added to the `IHederaTokenService` interface to support the updating the custom fees for fungible and non-fungible tokens.
+
+| Function Selector Hash |                                                                                        Function Signature                                                                                         |    Response    |                                 |
+|------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------|---------------------------------|
+| `0xe780c5d3`           | `function updateFungibleTokenCustomFees(address token,  IHederaTokenService.FixedFee[] memory fixedFees, IHederaTokenService.FractionalFee[] memory fractionalFees) returns (int64 responseCode)` | `ResponseCode` | The response code from the call |
+| `0x01f9eb7d`           | `function updateNonFungibleTokenCustomFees(address token, IHederaTokenService.FixedFee[] memory fixedFees, IHederaTokenService.RoyaltyFee[] memory royaltyFees) returns (int64 responseCode)`     | `ResponseCode` | The response code from the call |
+
+### Smart Contract Service Module
+
+- UpdateTokenCustomFeesTranslator - This class will be responsible for handling the `updateFungibleTokenCustomFees` and `updateNonFungibleTokenCustomFees` selectors and dispatching them to the corresponding HAPI calls.
+- UpdateTokenCustomFeesDecoder - This class provides methods  and constants for decoding the given `HtsCallAttempt` into a `TransactionBody` for `TokenFeeScheduleUpdate` call.
+
+## Security Implications
+
+The newly added flows will adopt the HAPI authorization logic. In this way the transaction would be successful only if the given contract making the call is set as a `feeScheduleKey` to the token or a threshold key including the calling contract ID is set for the same.
+We will apply the `TokenFeeScheduleUpdate` throttle mechanism.
+
+## Acceptance Tests
+
+### BDD Tests
```

Comment:
These are good tests - and good checklists.  But I wonder if it would be easier to make sense of in some kind of tabular format.  Some kind of arrangement where you can see what's going on is a cartesian product along different axes.  Or something like that.